backends/p4tools/testgen/lib/test_spec.cpp

```diff
@@ -0,0 +1,340 @@
+              "Variable is not a constant. It has type %1% instead. Has the test object %2% "
+              "been evaluated?",
+              value->type->node_type_name(), getObjectName());
+    return constant;
+}
+
+const Exact* Exact::evaluate(const Model& model) const {
+    const auto* evaluatedValue = model.evaluate(value);
+    return new Exact(getKey(), evaluatedValue);
+}
+
+cstring Exact::getObjectName() const { return "Exact"; }
+
+TableRule::TableRule(std::map<cstring, const FieldMatch> matches, int priority, ActionCall action,
+                     int ttl)
+    : matches(std::move(matches)), priority(priority), action(std::move(action)), ttl(ttl) {}
+
+const std::map<cstring, const FieldMatch>* TableRule::getMatches() const { return &matches; }
+
+int TableRule::getPriority() const { return priority; }
+
+const ActionCall* TableRule::getActionCall() const { return &action; }
+
+int TableRule::getTTL() const { return ttl; }
+
+cstring TableRule::getObjectName() const { return "TableRule"; }
+
+const TableRule* TableRule::evaluate(const Model& model) const {
+    std::map<cstring, const FieldMatch> evaluatedMatches;
+    for (const auto& matchTuple : matches) {
+        auto name = matchTuple.first;
+        auto match = matchTuple.second;
+        // This is a lambda function that applies the visitor to each variant.
+        const auto evaluatedMatch = boost::apply_visitor(
+            [model](auto const& obj) -> FieldMatch { return *obj.evaluate(model); }, match);
+        evaluatedMatches.insert({name, evaluatedMatch});
+    }
+    const auto* evaluatedAction = action.evaluate(model);
+    return new TableRule(evaluatedMatches, priority, *evaluatedAction, ttl);
+}
+
+const IR::P4Table* TableConfig::getTable() const { return table; }
+
+cstring TableConfig::getObjectName() const { return "TableConfig"; }
+
+TableConfig::TableConfig(const IR::P4Table* table, std::vector<TableRule> rules)
+    : table(table), rules(std::move(rules)) {}
+
+TableConfig::TableConfig(const IR::P4Table* table, std::vector<TableRule> rules,
+                         std::map<cstring, const TestObject*> tableProperties)
+    : table(table), rules(std::move(rules)), tableProperties(std::move(tableProperties)) {}
+
+const std::vector<TableRule>* TableConfig::getRules() const { return &rules; }
+
+const std::map<cstring, const TestObject*>* TableConfig::getProperties() const {
+    return &tableProperties;
+}
+
+const TestObject* TableConfig::getProperty(cstring propertyName, bool checked) const {
+    auto it = tableProperties.find(propertyName);
+    if (it != tableProperties.end()) {
+        return it->second;
+    }
+    if (checked) {
+        BUG("Unable to find test object with the label %1%. ", propertyName);
+    }
+    return nullptr;
+}
+
+void TableConfig::addTableProperty(cstring propertyName, const TestObject* property) {
+    tableProperties[propertyName] = property;
+}
+
+const TableConfig* TableConfig::evaluate(const Model& model) const {
```

Comment:
Based on the file name I would not have guessed that this file contains all these objects.